source/common/stats/tag_extractor_impl.cc

```diff
@@ -105,5 +116,50 @@ bool TagExtractorImpl::extractTag(absl::string_view stat_name, TagVector& tags,
   return false;
 }
 
+TagExtractorRe2Impl::TagExtractorRe2Impl(absl::string_view name, absl::string_view regex,
+                                         absl::string_view substr)
+    : TagExtractorImplBase(name, regex, substr), regex_(regex) {}
+
+bool TagExtractorRe2Impl::extractTag(absl::string_view stat_name, std::vector<Tag>& tags,
+                                     IntervalSet<size_t>& remove_characters) const {
+  PERF_OPERATION(perf);
+
+  if (substrMismatch(stat_name)) {
+    PERF_RECORD(perf, "re2-skip", name_);
+    return false;
+  }
+
+  // remove_subexpr is the first submatch. It represents the portion of the string to be removed.
+  re2::StringPiece remove_subexpr, value_subexpr;
+
+  // The regex must match and contain one or more subexpressions (all after the first are ignored).
+  if (re2::RE2::FullMatch(re2::StringPiece(stat_name.data(), stat_name.size()), regex_,
+                          &remove_subexpr, &value_subexpr) &&
+      !remove_subexpr.empty()) {
+
+    // value_subexpr is the optional second submatch. It is usually inside the first submatch
+    // (remove_subexpr) to allow the expression to strip off extra characters that should be removed
+    // from the string but also not necessary in the tag value ("." for example). If there is no
+    // second submatch, then the value_subexpr is the same as the remove_subexpr.
+    if (value_subexpr.empty()) {
+      value_subexpr = remove_subexpr;
+    }
+
+    tags.emplace_back();
+    Tag& tag = tags.back();
+    tag.name_ = name_;
+    tag.value_ = std::string(value_subexpr);
+
+    // Determines which characters to remove from stat_name to elide remove_subexpr.
+    std::string::size_type start = remove_subexpr.data() - stat_name.data();
+    std::string::size_type end = remove_subexpr.data() + remove_subexpr.size() - stat_name.data();
+    remove_characters.insert(start, end);
+    PERF_RECORD(perf, "re2-match", name_);
+    return true;
```

Comment:
Did you consider options for avoiding duplicating a lot of the logic/comments between here and the std::regex code?

Reply:
I messed around a bit and was only really able to profitably factor out the code to add a tag to the tags array. There were enough subtle difference in all the other parallels that it didn't seem worth it to refactor. LMK if you have other ideas.